src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -149,87 +169,158 @@ class FastSync(
         handleRequestFailure(assignedHandlers(ref), ref, "Unexpected error")
     }
 
-    private def removeRequestHandler(handler: ActorRef): Unit = {
-      context unwatch handler
-      assignedHandlers -= handler
+    def waitingForTargetBlockUpdate(state: FinalBlockProcessingResult): Receive = handleCommonMessages orElse {
+      case FastSyncTargetBlockSelector.Result(targetBlockHeader) =>
+        log.info(s"new target block with number ${targetBlockHeader.number} received")
+        if (targetBlockHeader.number >= syncState.targetBlock.number) {
+          updateTargetSyncState(state, targetBlockHeader)
+          syncState = syncState.copy(updatingTargetBlock = false)
+          context become this.receive
+          processSyncing()
+        } else {
+          syncState = syncState.copy(targetBlockUpdateFailures = syncState.targetBlockUpdateFailures + 1)
+          scheduler.scheduleOnce(syncRetryInterval, self, WaitForDownloadsToFinish(state))
+        }
+
+      case PersistSyncState => persistSyncState()
+
+      case WaitForDownloadsToFinish(sState) => updateTargetBlock(sState)
```

Comment:
isn't it already handled in `handleCommonMessages`?

Reply:
also naming seems to be a little misleading, you receive something that sounds like a command to wait: `WaitForDownloadsToFinish` but the action is updating target block. Why not just call this message `UpdateTargetBlock(state)`?